presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -700,62 +689,57 @@ protected RelationType visitSetOperation(SetOperation node, AnalysisContext cont
+        RelationType firstDescriptor = relationScopes.get(0).getRelationType().withOnlyVisibleFields();
         for (int i = 0; i < outputFieldTypes.length; i++) {
             Field oldField = firstDescriptor.getFieldByIndex(i);
             outputDescriptorFields[i] = new Field(oldField.getRelationAlias(), oldField.getName(), outputFieldTypes[i], oldField.isHidden());
         }
-        RelationType outputDescriptor = new RelationType(outputDescriptorFields);
-        analysis.setOutputDescriptor(node, outputDescriptor);
 
         for (int i = 0; i < node.getRelations().size(); i++) {
             Relation relation = node.getRelations().get(i);
-            RelationType descriptor = descriptors[i];
-            for (int j = 0; j < descriptor.getVisibleFields().size(); j++) {
+            Scope relationScope = relationScopes.get(i);
+            for (int j = 0; j < relationScope.getRelationType().getVisibleFields().size(); j++) {
                 Type outputFieldType = outputFieldTypes[j];
-                Type descFieldType = descriptor.getFieldByIndex(j).getType();
+                Type descFieldType = relationScope.getRelationType().getFieldByIndex(j).getType();
                 if (!outputFieldType.equals(descFieldType)) {
                     analysis.addRelationCoercion(relation, outputFieldTypes);
                     break;
                 }
             }
         }
-        return outputDescriptor;
+        return createScope(node, scope, outputDescriptorFields);
     }
 
     @Override
-    protected RelationType visitIntersect(Intersect node, AnalysisContext context)
+    protected Scope visitIntersect(Intersect node, Scope scope)
     {
         if (!node.isDistinct()) {
             throw new SemanticException(NOT_SUPPORTED, node, "INTERSECT ALL not yet implemented");
         }
 
-        return visitSetOperation(node, context);
+        return visitSetOperation(node, scope);
     }
 
     @Override
-    protected RelationType visitExcept(Except node, AnalysisContext context)
+    protected Scope visitExcept(Except node, Scope scope)
     {
         throw new SemanticException(NOT_SUPPORTED, node, "EXCEPT not yet implemented");
     }
 
     @Override
-    protected RelationType visitJoin(Join node, AnalysisContext context)
+    protected Scope visitJoin(Join node, Scope scope)
     {
         JoinCriteria criteria = node.getCriteria().orElse(null);
         if (criteria instanceof NaturalJoin) {
             throw new SemanticException(NOT_SUPPORTED, node, "Natural join not supported");
         }
 
-        AnalysisContext leftContext = new AnalysisContext(context, new RelationType());
-        RelationType left = process(node.getLeft(), context);
-        leftContext.setLateralTupleDescriptor(left);
-        RelationType right = process(node.getRight(), leftContext);
+        Scope left = process(node.getLeft(), scope);
+        Scope right = process(node.getRight(), left);
```

Comment:
`left` is also valid as parent scope for `JOIN UNNEST`. I was unable to model that differently than restoring `lateralScope` field in `Scope` class.